### ISSUE TITLE
pyfits/astropy backend is using deprecated functionality

### ISSUE THREAD

**Issue by DougBurke** [sherpa/sherpa#27]
# Summary

The pyfits back end uses deprecated functions and classes - it includes `new_table`, `CardList`, and `append` but I have not done an exhaustive check. Once #6 is resolved we should address these issues.

The error, whereby the Chandra PHA file can not be written out is discussed in #46.

# Details

The following test script:

```
% cat test_fits.py
# does writing out a FITS file cause warnings?
import sherpa
from sherpa.astro import ui
print("*** Sherpa version: {}".format(sherpa.__version__))
ui.load_arrays(1, [1,2,3], [4,5,6])
ui.save_data(1, 'test_fits.dat', ascii=True, clobber=True)
print("*** created test_fits.dat ***")
ui.save_data(1, 'test_fits.fits', ascii=False, clobber=True)
print("*** created test_fits.fits ***")
```

causes the following messages when writing the FITS version of the data. 

## When using pyfits

I meant to use the master branch, but used PR #26 by accident. As the code changes there are not functional, it is not the cause of the error (although the line numbering may be out slightly).

```
% conda list pyfits
# packages in environment at /home/naridge/local/anaconda/envs/sherpa-pep8:
#
You are using pip version 6.1.1, however version 7.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
pyfits                    3.3.0                np18py27_1  

% python test_fits.py 
WARNING: imaging routines will not be available, 
failed to import sherpa.image.ds9_backend due to 
'RuntimeErr: DS9Win unusable: Could not find ds9 on your PATH'
WARNING: failed to import sherpa.astro.xspec; XSPEC models will not be available
*** Sherpa version: 4.7+375.g1693f34
*** created test_fits.dat ***
/home/naridge/local/anaconda/envs/sherpa-pep8/lib/python2.7/site-packages/sherpa/astro/io/pyfits_backend.py:877: PyfitsDeprecationWarning: The new_table function is deprecated as of version 3.3 and may be removed in a future version.

        Use :meth:`BinTableHDU.from_columns` for new BINARY tables or :meth:`TableHDU.from_columns` for new ASCII tables instead.
  tbl = pyfits.new_table(pyfits.ColDefs(collist))
*** created test_fits.fits ***
```

## When using astropy

This is using PR #6 

```
% conda list astropy
# packages in environment at /home/naridge/local/anaconda/envs/sherpa-astropy:
#
You are using pip version 6.1.1, however version 7.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
astropy                   1.0.2                np19py27_0  

% python test_fits.py 
WARNING: imaging routines will not be available, 
failed to import sherpa.image.ds9_backend due to 
'RuntimeErr: DS9Win unusable: Could not find ds9 on your PATH'
WARNING: failed to import sherpa.astro.xspec; XSPEC models will not be available
*** Sherpa version: 4.7+2.gffe559b
*** created test_fits.dat ***
WARNING: AstropyDeprecationWarning: The new_table function is deprecated and may be removed in a future version.
        Use :meth:`BinTableHDU.from_columns` for new BINARY tables or :meth:`TableHDU.from_columns` for new ASCII tables instead. [astropy.utils.decorators]
*** created test_fits.fits ***
```

## Other deprecation warnings

Note that I've seen other warnings from the pyfits back-end. For instance, if I load in a CIAO PHA file, then I get - this is from the `astropy` branch, but you see essentially the same from pyfits:

 - deprecation warning from the CardList class

 - deprecation warning from the append function

 - the error message is being tracked in a separate bug, #46

```
In [1]: from sherpa.astro import ui
WARNING: imaging routines will not be available, 
failed to import sherpa.image.ds9_backend due to 
'RuntimeErr: DS9Win unusable: Could not find ds9 on your PATH'
WARNING: failed to import sherpa.astro.xspec; XSPEC models will not be available

In [2]: ui.load_pha(1, 'grp.pi')
WARNING: [Errno 2] No such file or directory: 'srcs_src1.corr.arf'
WARNING: [Errno 2] No such file or directory: 'srcs_src1.rmf'

In [3]: ui.save_data('test.fits', ascii=False, clobber=True)
WARNING: AstropyDeprecationWarning: The CardList class has been deprecated; all its former functionality has been subsumed by the Header class, so CardList objects should not be directly created.  See the PyFITS 3.1.0 CHANGELOG for more details. [astropy.io.fits.card]
WARNING: AstropyDeprecationWarning: The append function is deprecated and may be removed in a future version.
        Use :meth:`Header.append` instead. [astropy.utils.decorators]
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-0a6e7bbb21bb> in <module>()
----> 1 ui.save_data('test.fits', ascii=False, clobber=True)

<string> in save_data(id, filename, bkg_id, ascii, clobber)

/home/naridge/local/anaconda/envs/sherpa-astropy/lib/python2.7/site-packages/sherpa/astro/ui/utils.pyc in save_data(self, id, filename, bkg_id, ascii, clobber)
   3215 
   3216         try:
-> 3217             sherpa.astro.io.write_pha(filename, d, ascii, clobber)
   3218         except IOErr:
   3219             try:

/home/naridge/local/anaconda/envs/sherpa-astropy/lib/python2.7/site-packages/sherpa/astro/io/__init__.pyc in write_pha(filename, dataset, ascii, clobber)
    398     data, col_names, hdr = _pack_pha( dataset )
    399     backend.set_pha_data(filename, data, col_names, hdr, ascii=ascii,
--> 400                          clobber=clobber)
    401 
    402 def pack_table(dataset):

/home/naridge/local/anaconda/envs/sherpa-astropy/lib/python2.7/site-packages/sherpa/astro/io/pyfits_backend.pyc in set_pha_data(filename, data, col_names, header, ascii, clobber, packup)
    898         if header[key] is None:
    899             continue
--> 900         hdrlist.append(pyfits.Card( str(key.upper()), header[key] ))
    901 
    902     collist = []

/home/naridge/local/anaconda/envs/sherpa-astropy/lib/python2.7/site-packages/astropy/io/fits/card.pyc in __init__(self, keyword, value, comment, **kwargs)
    446                 self.keyword = keyword
    447             if value is not None:
--> 448                 self.value = value
    449 
    450         if comment is not None:

/home/naridge/local/anaconda/envs/sherpa-astropy/lib/python2.7/site-packages/astropy/io/fits/card.pyc in value(self, value)
    568                                  (float, complex, bool, Undefined, np.floating,
    569                                   np.integer, np.complexfloating, np.bool_)):
--> 570             raise ValueError('Illegal value: %r.' % value)
    571 
    572         if isinstance(value, float) and (np.isnan(value) or np.isinf(value)):

ValueError: Illegal value:   This FITS file may contain long string keyword values that are
  continued over multiple keywords.  The HEASARC convention uses the &
  character at the end of each substring which is then continued
  on the next keyword which has the name CONTINUE.
specextract version 18 September 2014.

```

**Comment by cdeil** [comment]
@DougBurke - Just in case it's related for the AstropyDeprecationWarning from new_table (probably isn't), note: https://github.com/astropy/astropy/issues/3798

**Comment by DougBurke** [comment]
Thanks for the heads up, @cdeil.

Doug

**Comment by anetasie** [comment]
This error is present when I use pyfits (not astropy).

(sherpa-test)SampleFlux $ ipython
Python 2.7.10 |Continuum Analytics, Inc.| (default, May 28 2015, 17:04:42) 
Type "copyright", "credits" or "license" for more information.

IPython 3.1.0 -- An enhanced Interactive Python.
Anaconda is brought to you by Continuum Analytics.
Please check out: http://continuum.io/thanks and https://anaconda.org
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: from sherpa.astro import ui
WARNING: imaging routines will not be available, 
failed to import sherpa.image.ds9_backend due to 
'RuntimeErr: DS9Win unusable: Could not find ds9 on your PATH'
WARNING: failed to import sherpa.astro.xspec; XSPEC models will not be available

In [2]: ui.load_pha(1,'3c273.pi')
WARNING: systematic errors were not found in file '3c273.pi'
statistical errors were found in file '3c273.pi' 
but not used; to use them, re-read with use_errors=True
read ARF file 3c273.arf
read RMF file 3c273.rmf
WARNING: systematic errors were not found in file '3c273_bg.pi'
statistical errors were found in file '3c273_bg.pi' 
but not used; to use them, re-read with use_errors=True
read background file 3c273_bg.pi

In [3]: ui.save_data('test_save.fits')
/Users/aneta/Software/Anaconda2.0/anaconda/envs/sherpa-test/lib/python2.7/site-packages/pyfits/card.py:70: PyfitsDeprecationWarning: The CardList class has been deprecated; all its former functionality has been subsumed by the Header class, so CardList objects should not be directly created.  See the PyFITS 3.1.0 CHANGELOG for more details.
  PyfitsDeprecationWarning)
/Users/aneta/git/sherpa/sherpa/astro/io/pyfits_backend.py:896: PyfitsDeprecationWarning: The append function is deprecated as of version 3.1 and may be removed in a future version.

        Use :meth:`Header.append` instead.
  hdrlist.append(pyfits.Card( str(key.upper()), header[key] ))
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-af54e1278c10> in <module>()
----> 1 ui.save_data('test_save.fits')

<string> in save_data(id, filename, bkg_id, ascii, clobber)

/Users/aneta/git/sherpa/sherpa/astro/ui/utils.pyc in save_data(self, id, filename, bkg_id, ascii, clobber)
   4674 
   4675         try:
-> 4676             sherpa.astro.io.write_pha(filename, d, ascii, clobber)
   4677         except IOErr:
   4678             try:

/Users/aneta/git/sherpa/sherpa/astro/io/__init__.pyc in write_pha(filename, dataset, ascii, clobber)
    398     data, col_names, hdr = _pack_pha( dataset )
    399     backend.set_pha_data(filename, data, col_names, hdr, ascii=ascii,
--> 400                          clobber=clobber)
    401 
    402 def pack_table(dataset):

/Users/aneta/git/sherpa/sherpa/astro/io/pyfits_backend.pyc in set_pha_data(filename, data, col_names, header, ascii, clobber, packup)
    894         if header[key] is None:
    895             continue
--> 896         hdrlist.append(pyfits.Card( str(key.upper()), header[key] ))
    897 
    898     collist = []

/Users/aneta/Software/Anaconda2.0/anaconda/envs/sherpa-test/lib/python2.7/site-packages/pyfits/card.pyc in __init__(self, keyword, value, comment, **kwargs)
    445                 self.keyword = keyword
    446             if value is not None:
--> 447                 self.value = value
    448 
    449         if comment is not None:

/Users/aneta/Software/Anaconda2.0/anaconda/envs/sherpa-test/lib/python2.7/site-packages/pyfits/card.pyc in value(self, value)
    567                                  (float, complex, bool, Undefined, np.floating,
    568                                   np.integer, np.complexfloating, np.bool_)):
--> 569             raise ValueError('Illegal value: %r.' % value)
    570 
    571         if isinstance(value, float) and (np.isnan(value) or np.isinf(value)):

ValueError: Illegal value:   This FITS file may contain long string keyword values that are
  continued over multiple keywords.  The HEASARC convention uses the &
  character at the end of each substring which is then continued
  on the next keyword which has the name CONTINUE..

**Comment by DougBurke** [comment]
@anetasie - this particular bug has been moved to #46 (the `ValueError` part). I've left this issue for the note about the use of deprecated functionality.

**Comment by DougBurke** [comment]
Fixed in PR #89

**Comment by DougBurke** [comment]
I have re-opened this because #89 was reverted. New development is in #116.
